src/runtime/transport.ts

```diff
@@ -0,0 +1,335 @@
+      case proto.EventType.message:
+        let channelMsg = toMessageEvent(this.encoding, msg.event);
+        this.handleChannelMessage(channelMsg);
+        return;
+      case proto.EventType.error:
+        let error = toErrorEvent(this.encoding, msg.event);
+        this.emit("error", error);
+        this.logger.error("recieved ", error);
+        return;
+      default:
+        throw new Error(`unknown message type ${msg.event_type}`);
+    }
+  }
+
+  clearHearbeat() {
+    clearTimeout(this.heartbeatId);
+  }
+  restartHeartbeat() {
+    this.clearHearbeat();
+    this.heartbeatId = setTimeout(
+      () => this.sendHeartbeat(),
+      this.config.heartbeatTimeout
+    );
+  }
+
+  send(msg: Uint8Array | string, saveOffline = false) {
+    if (this._connectionState !== "connected" && saveOffline) {
+      this.msgQueue.push(msg);
+    } else {
+      this.restartHeartbeat();
+      try {
+        this.ws.send(msg);
+      } catch (error) {
+        let message;
+        if (error instanceof Error) message = error.message;
+        else message = String(error);
+
+        if (/WebSocket is not open/.test(message) && saveOffline) {
+          this.msgQueue.push(msg);
+        }
+
+        this.logger.error(message);
+      }
+    }
+  }
+
+  sendHeartbeat() {
+    this.logger.debug("sending heartbeat");
+    this.restartHeartbeat();
+    this.ws.send(createHeartbeatEvent(this.encoding));
+  }
+
+  sendDisconnect() {
+    this.logger.debug("sending disconnect");
+    this.send(createDisconnectEvent(this.encoding));
+  }
+
+  connectionState(): ConnectionState {
+    return this._connectionState;
+  }
+
+  handleChannelMessage(msg: proto.MessageEvent) {
+    const channelState = this.channelsState.get(msg.channel);
+
+    if (!channelState) {
+      let error = {
+        code: 0,
+        message: `received msg for channel ${msg.channel} that doesn't exist`,
+      };
+      this.logger.error(error.message);
+      this.emit("error", error);
+      return;
+    }
+
+    channelState.position = msg.id;
+    channelState.listeners.forEach((listener) => listener(msg));
+  }
+
+  listen(channelName: string, listener: MessageEventListener) {
+    if (!this.channelsState.has(channelName)) {
+      this.channelsState.set(channelName, new ChannelState());
+    }
+
+    const channelState = this.channelsState.get(channelName);
+
+    if (!channelState) {
+      return;
+    }
+
+    channelState.listeners.push(listener);
+  }
+
+  attach(channelName: string) {
+    this.logger.debug("sending attach from ", channelName);
+    this.send(createAttachEvent(this.encoding, channelName));
+  }
+
+  detach(channelName: string) {
+    this.logger.debug("sending detach");
+    this.send(createDetachEvent(this.encoding, channelName));
+  }
+
+  subscribe(channelName: string, name: string, position: string) {
+    this.logger.debug("sending subscribe", channelName);
+    this.send(createSubscribeEvent(this.encoding, channelName, name, position));
```

Comment:
I forgot about it :) do we need a name in SubscribeEvent?

Reply:
No. The Subscribe is a bit confusing. From a user's perspective, a user subscribes to a channel and a `name` for a message. From the client and server perspective, the client subscribes to a channel and then will filter the messages based on the name of the message to the correct listeners.

Reply:
I've updated it to that now.